src/swupdd-main.c

```diff
@@ -22,751 +18,765 @@
+		ERR("Failed to fork: %s", strerror(errno));
+		return -errno;
+	}
+
+	close(fds[1]);
+	context->child = pid;
+	context->method = method;
+	sd_event *event = NULL;
+	r = sd_event_default(&event);
+	r = sd_event_add_io(event, NULL, fds[0], EPOLLIN, on_childs_output, context);
+	assert(r >= 0);
+	sd_event_unref(event);
+
+	return 0;
 }
 
-
-static gboolean _on_hash_dump (ClrSoftwareUpdate *object,
-                               GDBusMethodInvocation *invocation,
-                               GVariant *arg_options,
-                               const gchar *arg_filename,
-                               gpointer user_data)
+static int method_update(sd_bus_message *m,
+	                 void *userdata,
+	                 sd_bus_error *ret_error)
 {
-    daemon_state_t *dstate = (daemon_state_t *) user_data;
-    GError *error = NULL;
-    GList *args = NULL;
-    GVariantIter iter;
-    GVariant *value;
-    gchar *key;
-    gchar **argv;
-
-    g_message ("hashDump");
-    clr_software_update__complete_hash_dump (object, invocation,
-                                             (dstate->process == NULL));
-    if (dstate->process)
-        return FALSE;
-
-    args = g_list_append (args, g_strdup (SWUPD_CLIENT));
-    args = g_list_append (args, g_strdup ("hashdump"));
-
-    g_variant_iter_init (&iter, arg_options);
-    while (g_variant_iter_next (&iter, "{sv}", &key, &value))
-    {
-        _handle_common_options (key, value, args);
-        g_free (key);
-        g_variant_unref (value);
-    }
-
-    if (arg_filename)
-        args = g_list_append (args,
-                              g_strdup (arg_filename));
-
-    argv = _list_to_strv (args);
-    dstate->method = "hashDump";
-    dstate->process = g_subprocess_newv ((const gchar * const *) argv,
-                                         G_SUBPROCESS_FLAGS_STDOUT_PIPE |
-                                         G_SUBPROCESS_FLAGS_STDERR_MERGE,
-                                         &error);
-    if (dstate->process == NULL)
-    {
-        gchar *errmsg[2] = { NULL, NULL };
-        errmsg[0] = error->message;
-        clr_software_update__emit_request_completed (object,
-                                                     dstate->method,
-                                                     -1,
-                                                     (const gchar * const *) errmsg);
-        g_error_free (error);
-    }
-    else
-    {
-        g_subprocess_communicate_utf8_async (dstate->process,
-                                             NULL,
-                                             NULL,
-                                             _on_process_done,
-                                             dstate);
-    }
-
-    g_free (argv);
-    g_list_free_full (args, g_free);
-
-    return TRUE;
+	daemon_state_t *context = userdata;
+	int r = 0;
+	struct list *args = NULL;
+
+	if (context->child) {
+		r = -EAGAIN;
+		ERR("Busy with ongoing request to swupd");
```

Comment:
You can send this as an DBus error, composed of a code and message so it's more meaningful than just a boolean :-)